File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import defaultdict

def bfs(x, y, n, connect_info):
    visited = [0] * (n + 1)
    visited[x] = 1
    queue = [x]
    cnt = 1
    while queue:
        curr = queue.pop(0)
        for num in connect_info[curr]:
            # 방문한 적 없고 끊은 숫자가 아닌 경우
            if visited[num] == 0 and num != y:
                queue.append(num)
                visited[num] = 1
                cnt += 1
    return cnt

def solution(n, wires):
    # 연결된 전선 정보 저장
    connect_info = defaultdict(list)
    for x, y in wires:
        connect_info[x].append(y)
        connect_info[y].append(x)
    answer = 100
    # 하나씩 끊어보고 탐색 하기
    for x, y in wires:
        value1 = bfs(x, y, n, connect_info)
        value2 = n - value1
        value = abs(value1 - value2)
        if value < answer:
            answer = value 
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function bfs(x, y, n, connectInfo) {
  const visited = Array(n + 1).fill(0)
  visited[x] = 1
  const queue = [x]
  let cnt = 1
  // 큐가 있는 경우에만
  while (queue.length > 0) {
    const curr = queue.shift()
    for (const num of connectInfo[curr]) {
      // 방문한적 없고 끊은 전력망이 아닌경우에만
      if (visited[num] === 0 && num !== y) {
        queue.push(num)
        visited[num] = 1
        cnt += 1
      }
    }
  }
  return cnt
}

function solution(n, wires) {
  // 얕은 복사하는법 (처음 배운 것)
  const connectInfo = Array.from({ length: 101 }, () => []);
  for (const wire of wires) {
    connectInfo[wire[0]].push(wire[1])
    connectInfo[wire[1]].push(wire[0])
  }
  let answer = 100
  for (const wire of wires) {
    const value1 = bfs(wire[0], wire[1], n, connectInfo)
    const value2 = n - value1
    const value = Math.abs(value1- value2)
    if (value < answer) {
      answer = value
    }
  }
  return answer
}